clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -562,8 +562,18 @@ private void validate(String name, Map<String, Object> parsed, Map<String, Confi
                     }
                 }
                 config.recommendedValues(recommendedValues);
-                if (value != null && !recommendedValues.isEmpty() && !recommendedValues.contains(value)) {
-                    config.addErrorMessage("Invalid value for configuration " + key.name);
+                if (value != null && !recommendedValues.isEmpty()) {
+                    // Special handling of LIST type here, we need to make sure
+                    // every value in the list is valid.
+                    if (value instanceof List<?>) {
```

Comment:
This seems inconsistent with how we handle validation of values during normal parsing. We don't special case `LIST` there.

Why do we need to special case `List` here? I get that it makes it easier to validate something about each entry, but that also means you can't validate the entire list (e.g. you couldn't write something like `NonEmptyListValidator`). If we wanted per-entry validation, could we instead implement a `class ListValidator implements Validator` that accepts another `Validator` and runs it on each entry?